---
title: single cell
---

# single cell annotation

please go to: [https://sci-hub.mksa.top/10.1038/s41596-021-00534-0]


# single cell technology

* 10X


* SMART-seq


# analysis tools:

    * R

        - Seurat

        - Monocle

    * Python

        - scanpy

        - svitools



# Cellranger

https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/tutorial_ov

https://support.10xgenomics.com/single-cell-gene-expression/software/downloads/latest

install cellranger

PATH=$PATH:/home/keke/app/cellranger-7.1.0/bin

intall bcl2fastq2

https://support.illumina.com/downloads/bcl2fastq-conversion-software-v2-20.html

mamba  install -c dranew bcl2fastq

get reference

curl -O https://cf.10xgenomics.com/supp/cell-exp/refdata-gex-GRCh38-2020-A.tar.gz


## example

1. Download example data

    ```
    curl -O https://cf.10xgenomics.com/supp/cell-exp/cellranger-tiny-bcl-1.2.0.tar.gz
    curl -O https://cf.10xgenomics.com/supp/cell-exp/cellranger-tiny-bcl-simple-1.2.0.csv
    tar -zxvf cellranger-tiny-bcl-1.2.0.tar.gz
    ```
2. mdkfastq

    ```
    cellranger mkfastq --id=tutorial_walk_through \
        --run=cellranger-tiny-bcl-1.2.0 \
        --csv=cellranger-tiny-bcl-simple-1.2.0.csv
    ```

3. count

    ```
    curl -O  https://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_1k_v3/pbmc_1k_v3_fastqs.tar
    cellranger count --id=run_count_1kpbmcs \
        --fastqs=pbmc_1k_v3_fastqs \
        --sample=pbmc_1k_v3 \
        --transcriptome=/home/keke/data2/ref/refdata-gex-GRCh38-2020-A\
        --noexit
    ```


4 Q&A


    1. fastq-input for cellranger?

        * https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/fastq-input

    2. How to do when facing erros?

        * google

        * re-run

        * re-boot

        * pray to god