<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.1.189">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Bioinfor - single cell</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Bioinfor</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html">Home</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="../../qmd/01_bioinformatics/01_index.html" aria-current="page">Bioinformatics</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../qmd/02_pipeline/01_index.html">Pipeline</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../qmd/03_ml/01_index.html">ML</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../qmd/04_tools/01_index.html">Tools</a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/SQiangqiang"><i class="bi bi-twitter" role="img">
</i> 
 </a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/Jinjin89"><i class="bi bi-github" role="img" aria-label="github">
</i> 
 </a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://linkedin.com/in/keke-sun-6a9371267"><i class="bi bi-linkedin" role="img">
</i> 
 </a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">single cell</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../qmd/01_bioinformatics/01_index.html" class="sidebar-item-text sidebar-link">Bioinformatics</a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">database</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../qmd/01_bioinformatics/11_ncbi.html" class="sidebar-item-text sidebar-link">NCBI</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../qmd/01_bioinformatics/12_ensembl.html" class="sidebar-item-text sidebar-link">Ensembl</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../qmd/01_bioinformatics/13_genome.html" class="sidebar-item-text sidebar-link">Neocleotide</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../qmd/01_bioinformatics/14_tcga.html" class="sidebar-item-text sidebar-link">TCGA</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">sc</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../qmd/01_bioinformatics/21_sc.html" class="sidebar-item-text sidebar-link active">single cell</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">format</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../qmd/01_bioinformatics/31_data_format.html" class="sidebar-item-text sidebar-link">data format</a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../qmd/01_bioinformatics/41_omics.html" class="sidebar-item-text sidebar-link">Omics</a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#single-cell-annotation" id="toc-single-cell-annotation" class="nav-link active" data-scroll-target="#single-cell-annotation">single cell annotation</a></li>
  <li><a href="#single-cell-technology" id="toc-single-cell-technology" class="nav-link" data-scroll-target="#single-cell-technology">single cell technology</a></li>
  <li><a href="#analysis-tools" id="toc-analysis-tools" class="nav-link" data-scroll-target="#analysis-tools">analysis tools:</a></li>
  <li><a href="#cellranger" id="toc-cellranger" class="nav-link" data-scroll-target="#cellranger">Cellranger</a>
  <ul class="collapse">
  <li><a href="#example" id="toc-example" class="nav-link" data-scroll-target="#example">example</a></li>
  <li><a href="#snakemake" id="toc-snakemake" class="nav-link" data-scroll-target="#snakemake">snakemake</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">single cell</h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<section id="single-cell-annotation" class="level1">
<h1>single cell annotation</h1>
<p>please go to: [https://sci-hub.mksa.top/10.1038/s41596-021-00534-0]</p>
</section>
<section id="single-cell-technology" class="level1">
<h1>single cell technology</h1>
<ul>
<li><p>10X</p></li>
<li><p>SMART-seq</p></li>
</ul>
</section>
<section id="analysis-tools" class="level1">
<h1>analysis tools:</h1>
<pre><code>* R

    - Seurat

    - Monocle

* Python

    - scanpy

    - svitools</code></pre>
</section>
<section id="cellranger" class="level1">
<h1>Cellranger</h1>
<p>https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/tutorial_ov</p>
<p>https://support.10xgenomics.com/single-cell-gene-expression/software/downloads/latest</p>
<p>install cellranger</p>
<p>PATH=$PATH:/home/keke/app/cellranger-7.1.0/bin</p>
<p>intall bcl2fastq2</p>
<p>https://support.illumina.com/downloads/bcl2fastq-conversion-software-v2-20.html</p>
<p>mamba install -c dranew bcl2fastq</p>
<p>get reference</p>
<p>curl -O https://cf.10xgenomics.com/supp/cell-exp/refdata-gex-GRCh38-2020-A.tar.gz</p>
<section id="example" class="level2">
<h2 class="anchored" data-anchor-id="example">example</h2>
<ol type="1">
<li><p>Download example data</p>
<pre><code>curl -O https://cf.10xgenomics.com/supp/cell-exp/cellranger-tiny-bcl-1.2.0.tar.gz
curl -O https://cf.10xgenomics.com/supp/cell-exp/cellranger-tiny-bcl-simple-1.2.0.csv
tar -zxvf cellranger-tiny-bcl-1.2.0.tar.gz</code></pre></li>
<li><p>mdkfastq</p>
<pre><code>cellranger mkfastq --id=tutorial_walk_through \
    --run=cellranger-tiny-bcl-1.2.0 \
    --csv=cellranger-tiny-bcl-simple-1.2.0.csv</code></pre></li>
<li><p>count</p>
<pre><code>curl -O  https://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_1k_v3/pbmc_1k_v3_fastqs.tar
/home/keke/data2/app/cellranger-7.1.0/cellranger count \
    --id=run_count_1kpbmcs \
    --fastqs=pbmc_1k_v3_fastqs \
    --sample=pbmc_1k_v3 \
    --transcriptome=/home/keke/data2/ref/refdata-gex-GRCh38-2020-A\
    --noexit</code></pre></li>
<li><p>count on raw fastq file</p>
<p>suffix1： _S1_L001_R1_001.fastq.gz suffix2： _S1_L001_R2_001.fastq.gz</p>
<ul>
<li><p>test data1 zcat sample_name_S1_L001_R1_001.fastq.gz | head -n 100000 &gt; sample_1_S1_L001_R1_001.fastq</p>
<p>zcat sample_name_S1_L001_R2_001.fastq.gz | head -n 100000 &gt; sample_1_S1_L001_R2_001.fastq</p>
<p>gzip sample_1_S1_L001_R*fastq</p></li>
<li><p>test data2 ln -s /home/keke/base/raw/HRA003647/HRR858162/HRR858162_f1.fastq.gz sample_name_S1_L001_R1_001.fastq.gz ln -s /home/keke/base/raw/HRA003647/HRR858162/HRR858162_r2.fastq.gz sample_name_S1_L001_R2_001.fastq.gz</p></li>
</ul>
<pre><code>/home/keke/data2/app/cellranger-7.1.0/cellranger count \
    --id=out \
    --fastqs=sample_dir \
    --sample=sample_name \
    --transcriptome=/home/keke/data2/ref/refdata-gex-GRCh38-2020-A</code></pre></li>
</ol>
</section>
<section id="snakemake" class="level2">
<h2 class="anchored" data-anchor-id="snakemake">snakemake</h2>
<pre><code>import pandas as pd
import os
class sampleInfo:
    def __init__(self,input_file,samples_file_index = ["sample","reads"],file = "file_position"):
        # the input params
        self.input_file = input_file
        self.samples_file_index = samples_file_index
        self.sample_name = samples_file_index[0]
        self.sample_end = samples_file_index[1]
        self.file = file

    def run(self):
        if self.input_file.endswith("csv"):
            self.samples_df = pd.read_csv(self.input_file)
        elif self.input_file.endswith("txt") or self.input_file.endswith("tsv"):
            self.samples_df = pd.read_table(self.input_file,sep = "\s+")
        else:
            print("please supply sample metadata with csv/txt/tsv format!")
            exit("###############################\nThere is something wrong, you know, I only accept the files with csv/txt/tsv extension!\n###############################")

        # 1) get all the samples
        self.samples = self.samples_df[self.sample_name].unique()
        
        # 2) change all reads to upper cases
        self.samples_df[self.sample_end] = self.samples_df[self.sample_end].str.upper()

        # 3) set index
        self.samples_df.set_index(self.samples_file_index,inplace=True)
    
    def R1(self,wildcards):
        return self.samples_df.loc[wildcards.sample].at["R1",self.file]

    def R2(self,wildcards):
        return self.samples_df.loc[wildcards.sample].at["R2",self.file]
    def getSampleName(self,wildcards):
        return str(wildcards.sample)


#############################################
# load configs
#############################################

samples_info = sampleInfo("sample.csv")
# parse samples
samples_info.run()
samples = samples_info.samples
print(samples)

# prepare dirs
fastqs = "/home/keke/base/raw/processed/HRA003647/fastqs"
output = "/home/keke/base/raw/processed/HRA003647/output"


# suffix
suffix1 = "_S1_L001_R1_001.fastq.gz"
suffix2 = "_S1_L001_R2_001.fastq.gz"

#############################################
# Rules
#############################################

rule all:
    input:
        #expand(str(os.path.join(fastqs,"{sample}")) + suffix2,sample=samples)
        expand(str(os.path.join(output,"{sample}")),sample=samples)

rule step_1_createSoftLink:
    input:
        r1 = samples_info.R1,
        r2 = samples_info.R2
    output:
        fq1 = str(os.path.join(fastqs,"{sample}")) + suffix1,
        fq2 = str(os.path.join(fastqs,"{sample}")) + suffix2
    shell:
        """
        ln -s {input.r1} {output.fq1}
        ln -s {input.r2} {output.fq2}
        """

rule step_2_cellrangerCount:
    input: 
        fq1 = str(os.path.join(fastqs,"{sample}")) + suffix1,
        fq2 = str(os.path.join(fastqs,"{sample}")) + suffix2,
        fastqs = fastqs,
        output = output
    output:
        output_dir = str(os.path.join(output,"{sample}"))
    shell:
        """
        cd {input.output};
        /home/keke/data2/app/cellranger-7.1.0/cellranger count \
            --id={wildcards.sample} \
            --fastqs={input.fastqs} \
            --sample={wildcards.sample} \
            --transcriptome=/home/keke/data2/ref/refdata-gex-GRCh38-2020-A
        """</code></pre>
<p>4 Q&amp;A</p>
<pre><code>1. fastq-input for cellranger?

    * https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/fastq-input

2. How to do when facing erros?

    * google

    * re-run

    * re-boot

    * pray to god</code></pre>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>